src/csharp/Microsoft.Spark/Interop/Ipc/CallbackServer.cs

```diff
@@ -0,0 +1,235 @@
+        /// </summary>
+        private readonly ConcurrentDictionary<int, ICallbackHandler> _callbackHandlers =
+            new ConcurrentDictionary<int, ICallbackHandler>();
+
+        /// <summary>
+        /// Keeps track of all <see cref="CallbackConnection"/> objects identified by its
+        /// <see cref="CallbackConnection.ConnectionId"/>. The main thread creates a
+        /// <see cref="CallbackConnection"/> each time it receives a new socket connection
+        /// from the JVM side and inserts it into <see cref="_connections"/>. Each worker
+        /// task calls <see cref="CallbackConnection.Run"/> and removes the connection
+        /// once this call is finished. <see cref="CallbackConnection.Run"/> will not return
+        /// unless the <see cref="CallbackConnection"/> needs to be closed.
+        /// Also, <see cref="_connections"/> is used to bound the number of worker tasks
+        /// since it gives you the total number of active <see cref="CallbackConnection"/>s.
+        /// </summary>
+        private readonly ConcurrentDictionary<long, CallbackConnection> _connections =
+            new ConcurrentDictionary<long, CallbackConnection>();
+
+        /// <summary>
+        /// Each worker task picks up a CallbackConnection from _waitingConnections
+        /// and runs it.
+        /// </summary>
+        private readonly BlockingCollection<CallbackConnection> _waitingConnections =
+            new BlockingCollection<CallbackConnection>();
+
+        /// <summary>
+        /// Maintains a list of all the worker tasks.
+        /// </summary>
+        private readonly IList<Task> _workerTasks = new List<Task>();
+
+        /// <summary>
+        /// A <see cref="CancellationTokenSource"/> used to notify Tasks that operations
+        /// should be canceled.
+        /// </summary>
+        private readonly CancellationTokenSource _tokenSource = new CancellationTokenSource();
+
+        /// <summary>
+        /// Counter used to generate a unique id when registering a <see cref="ICallbackHandler"/>.
+        /// </summary>
+        private int _callbackCounter = 0;
+
+        private CallbackServer()
+        {
+            s_logger.LogInfo($"Starting CallbackServer.");
+            AppDomain.CurrentDomain.ProcessExit += (s, e) => Shutdown(s);
+            ISocketWrapper listener = SocketFactory.CreateSocket();
+            Run(listener);
+        }
+
+        internal static CallbackServer Instance => s_instance.Value;
+
+        /// <summary>
+        /// Produce a unique id and register a <see cref="ICallbackHandler"/> with it.
+        /// </summary>
+        /// <param name="callbackHandler">The handler to register.</param>
+        /// <returns>A unique id associated with the handler.</returns>
+        internal int RegisterCallback(ICallbackHandler callbackHandler)
+        {
+            int callbackId = Interlocked.Increment(ref _callbackCounter);
+            _callbackHandlers[callbackId] = callbackHandler;
```

Comment:
Looks like you can clean up this up if `Stop()`, `AwaitTermination()`, etc. are called on `StreamingQuery` by passing around callback id?

If so, this means we can create one `CallbackServer` per `ForEachBatch` call (by passing `CallbackServer` instance to `StreamingQuery` instead of callback id. Did you explore this approach? With this approach, you will just pass the callback server info directly to `callForeachBatch`, so no need to go thru `connectCallback`, etc. Wouldn't this approach simplify the code? It would be good to check pros/cons for these two approaches.

 How is PySpark handling such scenario (multiple foreach batch)?